satpy/etc/composites/viirs.yaml

```diff
@@ -685,3 +685,23 @@ composites:
       - name: I01
       - name: I03
     standard_name: cimss_cloud_type
+
+  cth:
+    description: >
+      VIIRS Cloud Top Height with colormap applied
+    compositor: !!python/name:satpy.composites.SingleBandCompositor
+    prerequisites:
+    - Cloud_Top_Height
+    standard_name: cth
+
+  viirs_cloud_top_height:
+    description: >
+      Cloud Top Height composite from NOAA/SNPP
+    compositor: !!python/name:satpy.composites.MaskingCompositor
+    prerequisites:
+      - Cloud_Top_Height
+      - cth
+    conditions:
+      - method: isnan
+        transparency: 100
+    standard_name: viirs_cloud_top_height
```

Comment:
Am I reading this correctly that both `Cloud_Top_Height` and `cth` come from the `viirs_l2` reader? What's the difference between them again?

I think this solution is reasonable, but we've also made it pretty common to include a boolean keyword argument in the file handler to enable or disable masking for certain products. This simplifies things in that users see the dataset as being part of the reader rather than being confused why something that comes from the file is being considered a composite. I think I'd prefer this, but I also don't know anything about the VIIRS L2 files and this/these products so I'm curious what you think?

Reply:
`cth` is just `Cloud_Top_Height` with the colormap applied, and then that is masked so the NaN values show up correctly with the applied colormap. The suggestion you made seems reasonable and if that's the best practice I can change it to have the boolean keyword.

Reply:
I don't know about "best" practice, but it is what I've been doing most of the time. What is the correct output you expect for NaN values going into a colormap? I guess I'm surprised that NaN in the dataset isn't coming out as fully transparent when saved as a geotiff.

You said that `cth` is the cloud top height with the colormap applied? How so? `cth` doesn't have an enhancement applied from what I see. Otherwise, if you meant that `Cloud_Top_Height` is the one with the colormap applied, that doesn't happen until after this compositing step unless I'm missing something. That is, MaskingCompositor and SingleBandCompositor don't apply enhancements internally. So the intention with this composite definition is to take a single band (L) array and add an Alpha (LA) where it is transparent where the data has NaN values. Isn't that already the default behavior?

Reply:
That is what I thought the default behavior would be. For whatever reason NaN values were being shown as the first color in my palette and were not transparent, this composite fixed that problem for me.

Reply:
Before I look into this more, do you *need* the palette to be preserved in your output or would an RGB/A as output be OK with you?

Basically, there is the colorized and the palettized variants of applying a colormap to something. I forget what the default behavior of the geotiff writer is, but there is an option to say whether the final dataset is saved with a geotiff ColorTable or converted to an RGB/A first. I think by default, especially since your colormap doesn't have an Alpha channel, the palettize enhancement will try to assign every incoming value to some color in that colormap. If you use the colorize enhancement instead I think it provides some freedom to the enhancement step to say "this is NaN, that's invalid, don't assign it a color in the colormap".

Reply:
The reason I chose the palettize instead of colorize in this case is because (if I'm not mistaken) colorize will interpolate between the color values given. E.g if I have a 900m and 950m color and a value of 925m at a given point it will interpolate between the two, when it should just be the 900m color. I am hoping to avoid that interpolation. Colorizing instead of using a palette did mask the NaN by default for me (hence why I don't have composites for the other datasets), like the comment you linked suggested.

Reply:
I have to get some other work done, but do you have a screenshot of what the CTH colormap is supposed to look like? What we've done in the past for interpolation issues is replicated the colors so you have one block of color from 0 to 5 and then the next color is 5.0001 to 10.0, then the next is 10.0001 to 15.0, and so on. Not great, but it allows us to cheat in most cases without having to have special cases for discrete colormaps.

Skimming the colors in the CTH colormap, I'm curious how you generated it. In the first block of colors I see the G and the B hover between 0 and 2, is that difference even perceptible by humans?

Reply:
As you can see in the screenshot it looks like there are only ~15 colors but in the json worldview uses there are many similar colors within each box. I reflected that json in my colormap, but I could just simplify it by using one color per "box" in the image and not use a palette to avoid headaches.
<img width="277" alt="Screenshot 2024-02-08 at 10 17 54 AM" src="https://github.com/pytroll/satpy/assets/153020597/d99729fa-ef9d-4356-8332-de6323e5b8d1">

Reply:
My current recommendation then is to do what I suggested above. You would use colorize and you would define your colormap as color1 -> color1 with values (control points) A to B, then color2 -> color2 with values B + 0.0001 + C, and so on. If you only defined the colors with a single point then you'd get interpolation between them. This is what I've ended up doing and I'm not sure there are many (any) other options within Satpy at the moment for a discrete colormap like this.

Reply:
This should be updated now. I removed the enhancements and switched from palette to colormap with the values and it produced the same result.